### ISSUE TITLE
Amazon.it - refund column always N/A

### ISSUE THREAD

**Issue by antwal** [philipmulcahy/azad#69]

**Comment by ScottMcNay** [comment]
@antwal, is this working now?

**Comment by galarosa** [comment]
Still not working for me. 
I'm not a developer but, as stated in the comment at line 302 of order.ts a different field name have to be searched or better an extraction by regex on strings like "Totale rimborso".

**Comment by philipmulcahy** [comment]
@galarosa, @antwal I pushed 1.9.12 last night which restores the ability to save debugging dumps of specific amazon orders from your account - please see the instructions on https://github.com/philipmulcahy/azad/blob/master/README.md#how-to-save-a-debug-file and https://github.com/philipmulcahy/azad/issues/98#issuecomment-653813986 and let me know here when you've done so.

**Comment by galarosa** [comment]
I've tried but it doesn't seems to create the debugging dump. In only show in the console this string: "dumpOrderDiagnostics: 405-XXXXXXX-XXX..... "

**Comment by antwal** [comment]
@philipmulcahy I am currently not using the extension as I have made a desktop version of it for my needs; but I can perform a test as I still keep it installed.

**Comment by philipmulcahy** [comment]
@galarosa have you got 1.9.12? 1.9.11 had the behaviour you're describing.

**Comment by galarosa** [comment]
Yes, is any way to enable additional messages in the console?

**Comment by galarosa** [comment]
This is the Google Chrome's release "Versione 84.0.4147.89 (Build ufficiale) (a 64 bit)" on windows 10.
And this is what is reported in extension details:
"Amazon Order History Reporter
On
Descrizione
This extension extracts order history from your Amazon account.
Versione
1.9.12
Dimensioni
62,6 MB
Autorizzazioni"
could it be a problem that Chrome is in italian?

**Comment by antwal** [comment]
hi, 

yes not working

my first order line: 
![order-line](https://user-images.githubusercontent.com/410736/88279517-7d98d800-cce4-11ea-9a94-1541ebc42309.png)

my console debug:
![order-console-2](https://user-images.githubusercontent.com/410736/88279567-986b4c80-cce4-11ea-92dc-33111bf014cd.png)

all requests are made as "language=en_GB" 

some important parts in log:

```
12:06:48.558 request_scheduler.ts:176 Queuing https://www.amazon.it/gp/css/order-history?opt=ab&digitalOrders=1&unifiedOrders=1&returnTo=&orderFilter=year-2020&startIndex=60&language=en_GB with 1
12:07:48.467 [Violation] 'load' handler took <N>ms
12:07:48.467 [Violation] 'load' handler took <N>ms
12:07:48.467 [Violation] 'load' handler took <N>ms
12:07:48.467 [Violation] 'load' handler took <N>ms
12:07:48.467 [Violation] 'load' handler took <N>ms
12:07:48.467 [Violation] 'load' handler took <N>ms
12:07:48.467 [Violation] 'load' handler took <N>ms
12:07:48.467 [Violation] 'load' handler took <N>ms
12:06:48.560 request_scheduler.ts:222 Executing https://www.amazon.it/gp/css/order-history?opt=ab&digitalOrders=1&unifiedOrders=1&returnTo=&orderFilter=year-2020&startIndex=50&language=en_GB with queue size 1 and priority 2
```

```
12:07:49.721 request_scheduler.ts:222 Executing https://www.amazon.it/gp/your-account/order-details/ref=oh_aui_or_o01_?ie=UTF8&orderID=402-2320689-0398744 with queue size 95 and priority 402-2320689-0398744
12:07:49.729 date.ts:67 could not parse date: Ordinato il 14 luglio 2020
normalizeDateString @ date.ts:66
order_date @ order.ts:81
extractDetailFromDoc @ order.ts:330
event_converter @ order.ts:356
protected_converter @ request_scheduler.ts:241
(anonymous) @ request_scheduler.ts:334
load (async)
_sendOne @ request_scheduler.ts:298
_execute @ request_scheduler.ts:254
_executeSomeIfPossible @ request_scheduler.ts:352
(anonymous) @ request_scheduler.ts:333
load (async)
_sendOne @ request_scheduler.ts:298
_execute @ request_scheduler.ts:254
_executeSomeIfPossible @ request_scheduler.ts:352
schedule @ request_scheduler.ts:183
receiveOrdersCount @ order.ts:677
protected_callback @ request_scheduler.ts:232
(anonymous) @ request_scheduler.ts:338
load (async)
_sendOne @ request_scheduler.ts:298
_execute @ request_scheduler.ts:254
_executeSomeIfPossible @ request_scheduler.ts:352
schedule @ request_scheduler.ts:183
sendGetOrderCount @ order.ts:613
(anonymous) @ order.ts:740
getOrdersForYearAndQueryTemplate @ order.ts:700
(anonymous) @ order.ts:904
fetchYear @ order.ts:904
(anonymous) @ order.ts:939
getOrdersByYear @ order.ts:935
(anonymous) @ inject.ts:91
Promise.then (async)
fetchAndShowOrders @ inject.ts:91
(anonymous) @ inject.ts:132
12:07:49.876 request_scheduler.ts:331 Finished https://www.amazon.it/gp/css/order-history?opt=ab&digitalOrders=1&unifiedOrders=1&returnTo=&orderFilter=year-2020&startIndex=50&language=en_GB with queue size 95
```

I try to download the last commit and understand where the problem is, most likely it is a language problem.

**Comment by philipmulcahy** [comment]
In my last post, I was asking (with insufficient clarity) "does 1.9.12 give you the ability to download dump files?" using the right click feature?

1.9.12 definitely doesn't know any more about Italian. The reason I want the dump files is so I can avoid sending you many experimental versions of the extension as we work out the fix to the parsing problem.
<img width="561" alt="Screenshot 2020-07-23 at 20 36 35" src="https://user-images.githubusercontent.com/1668892/88331128-65e14400-cd24-11ea-9035-3fda2b8cf805.png">

If you want to fix the problem yourself, that's great (really!), but a pull request will need to be accompanied by order data in a dump file so as to enable an appropriate test.

**Comment by antwal** [comment]
@philipmulcahy since all sensitive data, such as address, credit card, etc. they are included in the debug file, it would be better to avoid publishing them on github; if in the debug file you had foreseen the possibility of change this data with fictitious values, it wouldn't have been a problem to publish them.

**Comment by antwal** [comment]
@philipmulcahy you need change this line for italian refund:

https://github.com/philipmulcahy/azad/blob/9583e14982b3c57b6240fe43cd6e77537d5bc124/src/js/order.ts#L302

with:

`['Totale rimborso'].map( //TODO other field names?`

then refund column working:

![refund](https://user-images.githubusercontent.com/410736/88346633-1b25f300-cd49-11ea-83b5-e53df0737921.png)

@galarosa @ScottMcNay hope now is fixed this problem

**Comment by philipmulcahy** [comment]
Debug files are not published.

**Comment by philipmulcahy** [comment]
I/We considered scrubbing PII a while back, but decided not to do it for three reasons:

1) It gives a false sense of security - how would we know it worked and continued to work (without you handing over all your sensitive details and having them in a meta test pack) - for this reason I wouldn’t be happy with publishing the scrubbed files.
2) Modifying the files increases the risk that we create invalid tests - e.g. end up getting test passes for things that are really broken.
3) It’s a lot of work (I’ve worked on scrubbing systems)

Only I have access to the project’s gmail account: azadextension@gmail.com <mailto:azadextension@gmail.com> and only one other developer has access to the github repo that contains the dumps we have collected.

Yours,

Philip

**Comment by galarosa** [comment]
> with:
> 
> `['Totale rimborso'].map( //TODO other field names?`
> 
> then refund column working:

With this change I get values in the refund column but the export to csv functionality stopped.

In the console I get this errors during the table preparation:
```
569settings.ts:84 JSON.parse blew up with: SyntaxError: Unexpected token u in JSON at position 0 while parsing:undefined
(anonymous) @ settings.ts:84
```
and this ones trying to download the csv:
```
table.ts:361 Uncaught (in promise) TypeError: cell_done_promises is not iterable (cannot read property undefined)
    at table.ts:361
    at Array.forEach (<anonymous>)
    at table.ts:360
(anonymous) @ table.ts:361
(anonymous) @ table.ts:360
Promise.then (async)
reallyDisplayOrders @ table.ts:385
(anonymous) @ table.ts:495
Promise.then (async)
displayOrders @ table.ts:493
(anonymous) @ table.ts:475
table.ts:361 Uncaught (in promise) TypeError: cell_done_promises is not iterable (cannot read property undefined)
    at table.ts:361
    at Array.forEach (<anonymous>)
    at table.ts:360
(anonymous) @ table.ts:361
(anonymous) @ table.ts:360
Promise.then (async)
(anonymous) @ table.ts:475
504settings.ts:84 JSON.parse blew up with: SyntaxError: Unexpected token u in JSON at position 0 while parsing:undefined
(anonymous) @ settings.ts:84
```

**Comment by antwal** [comment]
@galarosa Yes I have not tried the export or other things, unfortunately I don't know well typescript if instead it was written in nodejs I had no problems making the necessary changes. You must wait for @philipmulcahy to make the necessary changes to integrate support for the various languages

**Comment by antwal** [comment]
@galarosa the error:

`504settings.ts:84 JSON.parse blew up with: SyntaxError: Unexpected token u in JSON at position 0 while parsing:undefined`

is generate from `azad_settings` which returns an undefined value

me the only errors it currently generates are:

`JSON.parse blew up with: SyntaxError: Unexpected token u in JSON at position 0 while parsing:undefined`

and 

`date.ts:67 could not parse date: Ordinato il 12 aprile 2020`

here on chrome extensions errors:

![image](https://user-images.githubusercontent.com/410736/88390091-02085b00-cdb8-11ea-92f0-8043426af91d.png)

only if i download csv do i see the error you say

![image](https://user-images.githubusercontent.com/410736/88390329-6b886980-cdb8-11ea-8b28-4b190e35c1e5.png)

if instead you put the check on the checkbox and use "Refund" instead of "Totale rimborso" the error occurs equally when you download CSV or at least in my case:

![image](https://user-images.githubusercontent.com/410736/88390450-9ecaf880-cdb8-11ea-896e-8dfadceb9cbd.png)

error here:

![image](https://user-images.githubusercontent.com/410736/88390888-6d066180-cdb9-11ea-92f3-c8cd6e515a38.png)

**Comment by antwal** [comment]
@philipmulcahy @galarosa tested working; change only

https://github.com/philipmulcahy/azad/blob/9583e14982b3c57b6240fe43cd6e77537d5bc124/src/js/order.ts#L302

with:

```
[ 'Refund', 'Totale rimborso'].map( //TODO other field names?
```

@galarosa for csv open another issue; this issue is fixed

**Comment by philipmulcahy** [comment]
I refer the honourable gentleman to my previous answer.

**Comment by philipmulcahy** [comment]
included 'Totale rimborso' in 7286b8a..b4bd21d - will be included in next release beyond 1.9.17.
